pkg/query/columnquery.go

```diff
@@ -142,9 +143,33 @@ func (q *ColumnQueryAPI) Query(ctx context.Context, req *pb.QueryRequest) (*pb.Q
 		return nil, err
 	}
 
+	if req.FilterQuery != nil {
+		p = filterProfileData(p, *req.FilterQuery)
+	}
+
 	return q.renderReport(ctx, p, req.GetReportType())
 }
 
+func filterProfileData(p *profile.Profile, filterQuery string) *profile.Profile {
+	filteredSamples := []*profile.SymbolizedSample{}
+	for _, s := range p.Samples {
+		var lines []profile.LocationLine
```

Comment:
We don't need to copy here. I would suggest the inside of this loop should be extracted to a separate function that's just called `keepSample` and if it returns true, then we append to `filteredSamples`. All the `keepSample` function then does is loop over all lines of all locations and on the first match it returns `true` and otherwise if no location line contained the search string, then it returns false.

Reply:
Done